src/net/ossl.cc

```diff
@@ -557,114 +610,237 @@ namespace tls {
+        case SSL_ERROR_SSL: {
+            auto ec = ERR_GET_REASON(ERR_peek_error());
+            if (ec == SSL_R_UNEXPECTED_EOF_WHILE_READING) {
+                // Probably shouldn't have during a write, but
+                // let's handle this gracefully
+                _eof = true;
+                return make_ready_future<stop_iteration>(stop_iteration::yes);
+            }
+            _error = make_exception_ptr(ossl_error::make_ossl_error(
+                "Error occurred during SSL write"));
+            // let's make sure there's no data to actually send
+            return wait_for_output().then_wrapped([this](auto f) {
+                try {
+                    f.get();
+                    return make_exception_future(_error);
+                } catch(...) {
+                    std::throw_with_nested(ossl_error::make_ossl_error(
+                        "Encountered unexpected error while handling SSL error during SSL write"));
                 }
-                auto bytes_written = SSL_write(_ssl.get(), ptr + off, size - off);
-                if(bytes_written <= 0){
-                    const auto ec = SSL_get_error(_ssl.get(), bytes_written);
-                    if (ec == SSL_ERROR_WANT_WRITE) {
-                        return pull_encrypted_and_send().then([]{
-                            return stop_iteration::no;
-                        });
-                    } else if (ec == SSL_ERROR_WANT_READ) {
-                        return do_get().then([](auto){
-                            return stop_iteration::no;
+            }).then([] {
+                return stop_iteration::no;
+            });
+        }
+        default:
+            // Some other unhandled situation
+            _error = std::make_exception_ptr(std::runtime_error(
+                "Unknown error encountered during SSL write"));
+            return make_exception_future<stop_iteration>(_error);
+        }
+    }
+
+    // Called post locking of the _out_sem
+    // This function takes and holds the sempahore units for _out_sem and
+    // will attempt to send the provided packet.  If a renegotiation is needed
+    // any unprocessed part of the packet is returned.
+    future<net::packet> do_put(net::packet p, semaphore_units<> units) {
+        if (!connected()) {
+            return make_ready_future<net::packet>(std::move(p));
+        }
+        assert(_output_pending.available());
+        return do_with(std::move(p), std::move(units), false,
+            [this](net::packet& p, semaphore_units<>& units, bool& renegotiate) {
+                // This do_until runs until either a renegotiation occurs or the packet is empty
+                return do_until(
+                    [this, &p, &renegotiate] { return eof() || renegotiate || p.len() == 0;},
+                    [this, &p, &renegotiate]() mutable {
+                        std::string_view frag_view =
+                            {p.fragments().begin()->base, p.fragments().begin()->size};
```

Comment:
NB: confirmed (for my own benefit I guess) that fragments are composed of signed chars.